job/service.go

```diff
@@ -738,3 +692,194 @@ func populateDownstreamDAGs(dagTree *tree.MultiRootTree, jobSpec models.JobSpec,
 
 	return rootNode, nil
 }
+
+// Refresh fetches all the requested jobs, resolves its dependencies, assign proper priority weights,
+// compile all jobs in the project and upload them to the destination store.
+func (srv *Service) Refresh(ctx context.Context, projectSpec models.ProjectSpec, namespaceJobNamePairs []models.NamespaceJobNamePair,
+	progressObserver progress.Observer) (err error) {
+	// Resolve dependency
+	if err = srv.resolveAndPersistDependency(ctx, projectSpec, namespaceJobNamePairs, progressObserver); err != nil {
+		// if err is caused by dependency resolution, ignore this as error.
+		var merrs *multierror.Error
+		if !errors.As(err, &merrs) {
+			return err
+		}
+	}
+
+	// Fetch dependency and enrich
+	jobDependencies, err := srv.dependencyResolver.FetchJobDependencies(ctx, projectSpec, progressObserver)
```

Comment:
Similarly here, as you are refetching the dependencies which were constructed above i believe your intention is to move to seperate place later. It would be better to do that right now.
1. FetchAllJobsAgain
2. FetchAllJobDependencies
3. EnrichJobsWithJobDependencies
4. EnrichJobsWithHookDependencies // This needs to be optimized by not making call to plugin inside enrichment rather we cache the dependencies between plugins during bootstrap and use pluginService for these kind of purposes.
5. PriorityResolution 
6. Grouping Namespecs // This step too, why are we keep on passing the projectJobSpecRep and fetching jobs lets just enrich the JobSpecModel with Project & NamespaceSpec or store ProjectId and NamespaceID in the JobSpec which can be used for grouping.

I expect all the data from datasources or third party integrations to happen only once and this data can be cached for optimization purposes. in here we are passing jobspecrepo in every place and fetching jobspecs again and again which can be avoided.

Reply:
updated. please help to recheck